gateway/models/guadrails.go

```diff
@@ -71,7 +139,110 @@ func DeleteGuardRailRules(orgID, ruleID string) error {
 	if _, err := GetGuardRailRules(orgID, ruleID); err != nil {
 		return err
 	}
+
+	// The associations will be deleted automatically due to the ON DELETE CASCADE
 	return DB.Table(tableGuardRails).
 		Where(`org_id = ? and id = ?`, orgID, ruleID).
 		Delete(&GuardRailRules{}).Error
 }
+
+// UpsertGuardRailRuleWithConnections creates or updates a guardrail rule and its connections in a single transaction
+func UpsertGuardRailRuleWithConnections(rule *GuardRailRules, connectionIDs []string, isNew bool) error {
+	// Clean empty connection IDs
+	var validConnectionIDs []string
+	for _, id := range connectionIDs {
+		if id != "" {
+			validConnectionIDs = append(validConnectionIDs, id)
+		}
+	}
+
+	return DB.Transaction(func(tx *gorm.DB) error {
+		// 1. Create or update the guardrail rule
+		var err error
+		if isNew {
+			// Create new guardrail rule
+			err = tx.Table(tableGuardRails).Model(rule).Create(rule).Error
+			if err == gorm.ErrDuplicatedKey {
+				return ErrAlreadyExists
+			} else if err != nil {
+				return err
+			}
+		} else {
+			// Update existing guardrail rule
+			res := tx.Table(tableGuardRails).
+				Model(rule).
+				Clauses(clause.Returning{}).
+				Updates(GuardRailRules{
+					Name:        rule.Name,
+					Description: rule.Description,
+					Input:       rule.Input,
+					Output:      rule.Output,
+					UpdatedAt:   rule.UpdatedAt,
+				}).
+				Where("org_id = ? AND id = ?", rule.OrgID, rule.ID)
+
+			if res.Error != nil {
+				return res.Error
+			}
+			if res.RowsAffected == 0 {
+				return ErrNotFound
+			}
+		}
+
+		// 2. Delete existing connections
+		if err := tx.Exec("DELETE FROM "+tableGuardRailsConnections+" WHERE org_id = ? AND rule_id = ?",
+			rule.OrgID, rule.ID).Error; err != nil {
+			return err
+		}
+
+		// 3. Add new connections
+		for _, connNameOrID := range validConnectionIDs {
```

Comment:
You're naming the variables as connection id in every route and now they are names or ids. Does the frontend need them to be names or ids? In case they are only ids, remove any logic related to names.